api/provider/aws/test_aws/service_read_test.go

```diff
@@ -0,0 +1,256 @@
+			DeployVersion: "1",
+			DesiredCount:  1,
+			PendingCount:  0,
+			RunningCount:  1,
+			Status:        ecs.DesiredStatusRunning,
+			Updated:       time.Time{},
+		},
+	}
+
+	expected := &models.Service{
+		Deployments:      expectedDeployments,
+		DesiredCount:     1,
+		EnvironmentID:    "env_id",
+		EnvironmentName:  "env_name",
+		LoadBalancerID:   "lb_id",
+		LoadBalancerName: "lb_name",
+		PendingCount:     0,
+		RunningCount:     1,
+		ServiceID:        "svc_id",
+		ServiceName:      "svc_name",
+	}
+
+	assert.Equal(t, expected, result)
+}
+
+func TestServiceRead_serviceNotFoundError(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+
+	mockAWS := awsc.NewMockClient(ctrl)
+	tagStore := tag.NewMemoryStore()
+	mockConfig := mock_config.NewMockAPIConfig(ctrl)
+
+	mockConfig.EXPECT().Instance().Return("test").AnyTimes()
+
+	tags := models.Tags{
+		{
+			EntityID:   "env_id",
+			EntityType: "environment",
+			Key:        "name",
+			Value:      "env_name",
+		},
+		{
+			EntityID:   "svc_id",
+			EntityType: "service",
+			Key:        "name",
+			Value:      "svc_name",
+		},
+		{
+			EntityID:   "svc_id",
+			EntityType: "service",
+			Key:        "environment_id",
+			Value:      "env_id",
+		},
+		{
+			EntityID:   "dpl_id",
+			EntityType: "deploy",
+			Key:        "arn",
+			Value:      "arn:aws:ecs:region:012345678910:task-definition/dpl_id:1",
+		},
+		{
+			EntityID:   "dpl_id",
+			EntityType: "deploy",
+			Key:        "name",
+			Value:      "dpl_name",
+		},
+		{
+			EntityID:   "dpl_id",
+			EntityType: "deploy",
+			Key:        "version",
+			Value:      "1",
+		},
+	}
+
+	for _, tag := range tags {
+		if err := tagStore.Insert(tag); err != nil {
+			t.Fatal(err)
+		}
+	}
+
+	describeServicesInput := &ecs.DescribeServicesInput{}
+	describeServicesInput.SetCluster("l0-test-env_id")
+	describeServicesInput.SetServices([]*string{aws.String("l0-test-svc_id")})
+
+	deployment := &ecs.Deployment{}
+	deployment.SetCreatedAt(time.Time{})
+	deployment.SetDesiredCount(int64(1))
+	deployment.SetPendingCount(int64(0))
+	deployment.SetRunningCount(int64(1))
+	deployment.SetStatus(ecs.DesiredStatusRunning)
+	deployment.SetTaskDefinition("arn:aws:ecs:region:012345678910:task-definition/dpl_id:1")
+	deployment.SetUpdatedAt(time.Time{})
+	deployments := []*ecs.Deployment{
+		deployment,
+	}
+
+	service := &ecs.Service{}
+	service.SetDeployments(deployments)
+	service.SetDesiredCount(int64(1))
+	service.SetPendingCount(int64(0))
+	service.SetRunningCount(int64(1))
+	services := []*ecs.Service{
+		service,
+	}
+
+	describeServicesOutput := &ecs.DescribeServicesOutput{}
+	describeServicesOutput.SetServices(services)
+
+	awsError := awserr.New("ServiceNotFoundException", "", nil)
+
+	mockAWS.ECS.EXPECT().
+		DescribeServices(describeServicesInput).
+		Return(&ecs.DescribeServicesOutput{}, awsError)
+
+	target := provider.NewServiceProvider(mockAWS.Client(), tagStore, mockConfig)
+	_, err := target.Read("svc_id")
```

Comment:
you didn't check this `err`.

also...

`IF _, ERR := FOO(): ERR != NIL { ... }`!!!

Reply:
@zpatrick In this test, I'm _expecting_ `Read()` to return an error, and I didn't do an inline check because I wanted to use`assert.Equal()`. The next two lines are where the error's checked:

```
target := provider.NewServiceProvider(mockAWS.Client(), tagStore, mockConfig)
_, err := target.Read("svc_id")

expectedErr := errors.Newf(errors.ServiceDoesNotExist, "Service 'l0-test-svc_id' does not exist")
assert.Equal(t, expectedErr, err)
```

Which is more important: using `assert.Equal()`, or using the inline `if`?

The inline `if` could look like this:
```
target := provider.NewServiceProvider(mockAWS.Client(), tagStore, mockConfig
expectedErr := errors.Newf(errors.ServiceDoesNotExist, "Service 'l0-test-svc_id' does not exist")
if _, err := target.Read("svc_id"); err != expectedErr {
    t.Fatalf("Error '%v' did not match expected error '%v'", err, expectedErr)
}
```

Reply:
oops totally missed that. However, I think we should only check the error status codes are equal, since we don't really care about the message. For example:
```
_, err := target.Read("svc_id")
assert.Equal(t, errors.ServiceDoesNotExist, err.(*errors.ServerError).ErrorCode)
```

Side note: this will panic if err is not of type `*errors.ServerError`. This could be improved by doing a `err, ok := err.(*errors.ServerError)` check

Reply:
✅